### ISSUE TITLE
Can't find migration version 128

### ISSUE THREAD

**Issue by azgul** [RocketChat/Docker.Official.Image#52]
It appears that my instance of [watchtower](https://github.com/v2tec/watchtower) has automagically updated my RocketChat instance to the latest version.

Now it wont launch because of a missing database migration.

```
Will load cache for users
5 records load from users
Will load cache for rocketchat_room
5 records load from rocketchat_room
Will load cache for rocketchat_subscription
14 records load from rocketchat_subscription
Will load cache for rocketchat_settings
720 records load from rocketchat_settings
Updating process.env.MAIL_URL
Will load cache for rocketchat_permissions
80 records load from rocketchat_permissions
Will load cache for rocketchat_roles
11 records load from rocketchat_roles
Starting Email Intercepter...
Warning: connect.session() MemoryStore is not
designed for a production environment, as it will leak
memory, and will not scale past a single process.
LocalStore: store created at
LocalStore: store created at
LocalStore: store created at
Setting default file store to GridFS
Thu, 21 Jun 2018 16:07:28 GMT connect deprecated multipart: use parser (multiparty, busboy, formidable) npm module instead at npm/node_modules/connect/lib/middleware/bodyParser.js:56:20
Thu, 21 Jun 2018 16:07:28 GMT connect deprecated limit: Restrict request size at location of read at npm/node_modules/connect/lib/middleware/multipart.js:86:15
Exception in callback of async function: Error: [Can't find migration version 128]
    at Object.Migrations._findIndexByVersion (/app/bundle/programs/server/packages/rocketchat_migrations.js:435:9)
    at Object.Migrations._migrateTo (/app/bundle/programs/server/packages/rocketchat_migrations.js:300:25)
    at Object.Migrations.migrateTo (/app/bundle/programs/server/packages/rocketchat_migrations.js:227:23)
    at xrun.js (/app/bundle/programs/server/app/app.js:7213:25)
    at fileEvaluate (packages/modules-runtime.js:343:9)
    at require (packages/modules-runtime.js:238:16)
    at /app/bundle/programs/server/app/app.js:12922:1
    at /app/bundle/programs/server/boot.js:411:36
    at Array.forEach (<anonymous>:null:null)
    at /app/bundle/programs/server/boot.js:220:19
    at /app/bundle/programs/server/boot.js:471:5
    at Function.run (/app/bundle/programs/server/profile.js:510:12)
    at /app/bundle/programs/server/boot.js:470:11
```

Can I run the migration manually somehow, and how do I avoid this happening in the future?

My `docker-compose.yml` for context:
```yaml
version: '2'

services:
  rocketchat:
    image: rocketchat/rocket.chat:latest
    restart: unless-stopped
    volumes:
      - ./uploads:/app/uploads
    environment:
      - PORT=3000
      - ROOT_URL=<snip>
      - MONGO_URL=mongodb://mongo:27017/rocketchat
      - MONGO_OPLOG_URL=mongodb://mongo:27017/local
    depends_on:
      - mongo
    ports:
      - 3000:3000
    networks:
    - proxy
    labels:
    - "traefik.enable=true"
    - "traefik.backend=rocketchat"
    - "traefik.frontend.rule=Host:<snip>"
    - "traefik.docker.network=reverseproxy_default"

  mongo:
    image: mongo:3.2
    restart: unless-stopped
    volumes:
     - ./data/db:/data/db
    command: mongod --smallfiles --oplogSize 128 --replSet rs0 --storageEngine=mmapv1
    labels:
      - "traefik.enable=false"
    networks:
    - proxy

networks:
  proxy:
    external:
      name: reverseproxy_default
```

**Comment by azgul** [comment]
I set the image version to `image: rocketchat/rocket.chat:0.66.0-rc.1` and voila:

```
{"line":"160","file":"rocketchat_migrations.js","message":"Migrations: Not migrating, already at version 128","time":{"$date":1529777900735},"level":"info"}
Using GridFS for custom sounds storage
Using GridFS for custom emoji storage
ufs: temp directory created at "/tmp/ufs"
Updating process.env.MAIL_URL
➔ System ➔ startup
➔ +----------------------------------------------------+
➔ |                   SERVER RUNNING                   |
➔ +----------------------------------------------------+
➔ |                                                    |
➔ |  Rocket.Chat Version: 0.66.0-rc.1                  |
➔ |       NodeJS Version: 8.9.3 - x64                  |
➔ |             Platform: linux                        |
➔ |         Process Port: 3000                         |
➔ |             Site URL: <snip>                       |
➔ |     ReplicaSet OpLog: Enabled                      |
➔ |          Commit Hash: b86c9333f6                   |
➔ |        Commit Branch: HEAD                         |
➔ |                                                    |
➔ +----------------------------------------------------+
```

I can only speculate, but perhaps `:latest` was updated twice. First to the release candidate then back to latest release (`0.65.2`?), effectively breaking the installs of everyone who auto updated.

**Comment by ChessSpider** [comment]
Hi @azgul its a risky fix you took; if there are DB changes between 0.66.0-RC.1 and 0.66.0-final then you won't get them. 

It seems they inadvertently pushed an update which updated the required db-level for the app, but did not include the migrations. So it should be possible to update the database to lower the required db-level. The ``migration`` collection in MongoDB to be precise. But I don't know what to set it to. It would have been nice to have some kind of official response from one of the devs, but alas.

The final release is due this Wednesday  [Final Release - On the 27th](https://github.com/RocketChat/Rocket.Chat/issues/11207). 

Azgul, were you using  https://hub.docker.com/r/rocketchat/rocket.chat/ or https://hub.docker.com/_/rocket.chat ? 

I am no longer going to use their images directly; instead I will use my private Docker repository to rebuild their Dockerfile once a week.. Should minimize the changes of it breaking due to their accidental pushes..

**Comment by azgul** [comment]
@ChessSpider Well, that doesn't sound good. I can't tell for certain what I _was_ using, but only what I am using now. As I didn't commit my changes to `docker-compose.yml`. 

> It seems they inadvertently pushed an update which updated the required db-level for the app, but did not include the migrations.

Interesting theory. I'll await some sort of official response and stick to rc1 until then.

**Comment by ChessSpider** [comment]
yeah.. luckily you went to a release-candidate so the chances of db-change are minimal, but still.

Anyway this prompted me to have another look at my continuous integration solution. I also use Watchtower. In addition to my previous comment, I also found this pull-request: https://github.com/v2tec/watchtower/pull/197. I am going to try to combine this functionality with some kind of incremental backup technology (e.g. something like http://www.admin-magazine.com/Articles/Using-rsync-for-Backups/(offset)/2 ). Then I can create a backup of the database before running the new version; which should allow easy downgrades in case something breaks again. 

I'll post my script here if everything works and you're interested.

**Comment by ChessSpider** [comment]
Oh, and additionally, it seems the images from https://hub.docker.com/_/rocket.chat may be a more stable option. E.g. they do not even include dev or rc candidates on that repository so hopefully mistakes like these are more harder to make.

Before, I was on https://hub.docker.com/r/rocketchat/rocket.chat/

**Comment by azgul** [comment]
Good point - I must've been on `rocketchat/rocket.chat` the entire time then.

**Comment by ChessSpider** [comment]
So my rocket.chat instance is working agian.

I created a new rocketchat instance, and logged into the mongo-database to find its current migration level.  

```
docker-compose exec mongo /bin/bash
root@373da192e035:/# mongo
MongoDB shell version: 3.2.20
connecting to: test
rs0:PRIMARY> use rocketchat
switched to db rocketchat
rs0:PRIMARY> db.migrations.find()
{ "_id" : "control", "version" : 121, "locked" : false, "lockedAt" : ISODate("2018-06-25T09:34:59.270Z"), "buildAt" : "2018-06-16T21:33:15.003Z" }
```

I then changed the migration-level of the broken rocketchat instance to 121:
```
 docker-compose exec mongo /bin/bash
mongo
rs0:PRIMARY> use rocketchat
switched to db rocketchat
drs0:PRIMARY>db.migrations.find()
{ "_id" : "control", "version" : 128, "locked" : true, "lockedAt" : ISODate("2018-06-21T21:24:50.824Z"), "buildAt" : "2018-06-16T21:33:15.003Z" }
rs0:PRIMARY> db.migrations.update({"_id": "control"}, {"version": 121, "locked": false})
WriteResult({ "nMatched" : 1, "nUpserted" : 0, "nModified" : 1 })
rs0:PRIMARY> db.migrations.find()
{ "_id" : "control", "version" : 121, "locked" : false }
exit
exit
docker-compose restart
```
Watch the logs, and it said ``{"line":"160","file":"rocketchat_migrations.js","message":"Migrations: Not migrating, already at version 121","time":{"$date":1529919716358},"level":"info"}``.

Now running: Rocket.Chat Version: 0.65.2   ~~ SAML login is still broken lol but at least existing sessions work again.

**Comment by ChessSpider** [comment]
Hi @azgul 

In case you're interested, I rebuild Watchtower with the pre-docker-command. You can find the Dockerfile I used to build it here:  https://github.com/v2tec/watchtower/pull/197#issuecomment-399940283

Then I used the following rsync backup script: https://github.com/pedroetb/rsync-incremental-backup/blob/master/rsync-incremental-backup-local  (slightly modified to not do the +r for everyone)

And extended the labes of the docker-compose rocketchat-service definitionto the following:
```
    labels:
      - "traefik.backend=rocketchat"
      - "traefik.frontend.rule=Host: your.domain.tld"
      - "com.centurylinklabs.watchtower.pre-update-command=/abs/path/to/rocketchat/rsync-incremental-backup-local"
```

**Comment by rodrigok** [comment]
@azgul we had a problem were our CI deployed our `0.66.0-rc.1` as the `latest` version, this happened for a few minutes and we removed the `latest` from our registry then we rebuild the correct latest version. You may dowloaded the latest in the middle of that process.

We are sorry about that problem. There are 2 ways to solve, one is accessing your database and at the collection `migrations` edit the only record you will see and set the field `version` to `121`.

The second option is exactly what you did, keep using the release candidate version, we are about to release the final version and will be available as `0.66.0` in a few hours.

That problem should not happen again, was caused by a step where we was updating how we deploy the images to docker hub.

I'll close this issue, please reopen it or let me know if you still have problems.

Thanks.

**Comment by janardna** [comment]
Exception in callback of async function: Error: [Can't find migration version 137]
